homeassistant/components/squeezebox/__init__.py

```diff
@@ -1 +1,64 @@
-"""The squeezebox component."""
+"""The Logitech Squeezebox integration."""
+
+import asyncio
+
+from pysqueezebox import async_discover
+
+from homeassistant.components.media_player import DOMAIN as MP_DOMAIN
+from homeassistant.config_entries import SOURCE_DISCOVERY, ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_PORT, EVENT_HOMEASSISTANT_START
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN
+from .media_player import _LOGGER
+
+DISCOVERY_TASK = "discovery_task"
+
+
+async def start_server_discovery(hass):
+    """Start a server discovery task."""
+
+    def _discovered_server(server):
+        asyncio.create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_DISCOVERY},
+                data={
+                    CONF_HOST: server.host,
+                    CONF_PORT: int(server.port),
+                    "uuid": server.uuid,
+                },
+            )
+        )
+
+    hass.data.setdefault(DOMAIN, {})
+    if DISCOVERY_TASK not in hass.data[DOMAIN]:
+        _LOGGER.debug("Adding server discovery task for squeezebox")
+        hass.data[DOMAIN][DISCOVERY_TASK] = hass.async_create_task(
+            async_discover(_discovered_server)
+        )
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Logitech Squeezebox component."""
+    if hass.is_running:
+        await start_server_discovery(hass)
+    else:
+        hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_START, start_server_discovery(hass)
+        )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Logitech Squeezebox from a config entry."""
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(entry, MP_DOMAIN)
+    )
+    return True
+
+
+async def async_unload_entry(hass, entry):
+    """Unload a config entry."""
+    return await hass.config_entries.async_forward_entry_unload(entry, MP_DOMAIN)
```

Comment:
Is discovery still going to be running if all entries are deleted?

Reply:
Yes. The server discovery will run if the integration is loaded. I think this behavior makes sense. For example, if the server `host`, `username`, or `password` changed, the user would delete the old entry and add a new one.

Reply:
Ideally we stop discovery if there are no more config entries loaded, and then restart it the next time a config entry it loaded.

This allows users to migrate from a squeezebox to some other device without restarting and having the overhead of discovery running for a device they no longer have.

Reply:
Ok. The discovery task will now end on unload of the last config entry.

Reply:
Thanks for fixing that.  We probably need to start it again if it’s not running when the first config entry is loaded instead of in async setup?

Reply:
You're right - we need to make sure it is running when we load a config entry. One nice thing about having it start in `async_setup` is that you can load the integration manually in `configuration.yaml` with `squeezebox:` and let it auto-discover squeezeboxes (which is a good substitute for using the deprecated `discovery` integration if `squeezebox` is the only thing left in `discovery` you care about).

I also realized the player discovery task was not cleaning up gracefully, so I fixed that too.